src/edu/csus/ecs/pc2/services/eventFeed/WebServer.java

```diff
@@ -107,80 +114,95 @@ public class WebServer implements UIPlugin {
+            store.load(null, null);
+            char[] password = keystorePassword.toCharArray();
+
+            KeyPair keyPair = generateKeyPair();
+            Certificate selfSign = selfSign(keyPair, "CN=pc2ef,O=PC^2,OU=PC^2");
+            
+            Certificate[] chain = new Certificate[1];
+            chain[0] = selfSign;
+            store.setKeyEntry("jetty", keyPair.getPrivate(), password, chain);
+
+            try (FileOutputStream fos = new FileOutputStream(PC2_KEYSTORE_FILE)) {
+                store.store(fos, password);
+                fos.close();
             }
-            String x500Name = "sun.security.x509" + ".X500Name";
-            Class<?> certKeyGenClass = Class.forName(certAndKeyGen);
-            Class<?> x500NameClass = Class.forName(x500Name);
-            Constructor<?> certKeyGenCons = certKeyGenClass.getConstructor(String.class, String.class);
-            Constructor<?> x500NameCons = x500NameClass.getConstructor(String.class);
-            Object keypair = certKeyGenCons.newInstance("RSA", "SHA256WithRSA");
-            String dn = "CN=pc2 jetty, OU=PC^2, O=PC^2, L=Unknown, ST=Unknown, C=Unknown";
-            Object subject = x500NameCons.newInstance(dn);
-            Method certAndKeyGenGenerate = certKeyGenClass.getMethod("generate", int.class);
-            certAndKeyGenGenerate.invoke(keypair, 2048);
-            Method certAndKeyGenPrivateKey = certKeyGenClass.getMethod("getPrivateKey");
-            PrivateKey rootPrivateKey = (PrivateKey) certAndKeyGenPrivateKey.invoke(keypair);
-            Method getSelfCertificate = certKeyGenClass.getMethod("getSelfCertificate", x500NameClass, long.class);
-
-            X509Certificate[] chain = new X509Certificate[1];
-            // create with a length of 1 (non-leap) year
-            long days = (long) 365 * 24 * 3600;
-            // Generate self signed certificate
-            chain[0] = (X509Certificate) getSelfCertificate.invoke(keypair, subject, days);
-
-            Principal issuer = chain[0].getSubjectDN();
-            String issuerSigAlg = chain[0].getSigAlgName();
-            byte[] inCertBytes = chain[0].getTBSCertificate();
-            X509CertInfo info = new X509CertInfo(inCertBytes);
-            info.set(X509CertInfo.ISSUER, (X500Name) issuer);
-
-            CertificateExtensions exts = new CertificateExtensions();
-            exts.set(SubjectKeyIdentifierExtension.NAME, new SubjectKeyIdentifierExtension(new KeyIdentifier(chain[0].getPublicKey()).getIdentifier()));
-            info.set(X509CertInfo.EXTENSIONS, exts);
-
-            X509CertImpl outCert = new X509CertImpl(info);
-            outCert.sign(rootPrivateKey, issuerSigAlg);
-
-            ks.setCertificateEntry("jetty", outCert);
-            ks.setKeyEntry("jetty", rootPrivateKey, keystorePassword.toCharArray(), chain);
         } catch (Exception ex) {
+            log.throwing("WebServer", "createKeyStoreAndFile", ex);
```

Comment:
If the EFClient is using a GUI (and I don't see how the WebServer can be started otherwise), I think the occurrence of an exception here should be displayed with a GUI dialog (the user might not be monitoring the console, and typically _won't_ be poring over logs).  

For example, the following block of code added between the above `log.throwing()` and the following `ex.printStackTrace()` would do the trick:
```java
            if (getController().isUsingGUI()) {
                String msg = "Warning: exception occurred during KeyStore creation: " + ex.getMessage();
                msg += "\nStack trace:";
                StackTraceElement[] stackTraceElements = ex.getStackTrace();
                for (StackTraceElement line : stackTraceElements) {
                    if (line.toString().contains("edu.csus.ecs.pc2")) {
                        msg += "\n" + line;
                    }
                }
                msg += "\n...";
                JOptionPane.showMessageDialog(null, msg, "Error creating WebServer KeyStore", JOptionPane.WARNING_MESSAGE);
            }

```

I've tested this code by forcing an exception above this `catch` clause and it works; here's an example:

![image](https://github.com/pc2ccs/pc2v9/assets/9256691/a2d76901-73d7-4701-b17f-0a53d1126871)

Reply:
The event feed CAN be (and for us in GNY is always) started in `--nogui `mode.  Page 140 of the _PC2 Administrators Guide_ explains it in Appendix M.  In this case, the `pc2ws.properties` file is examined for the settings.  In addition, the https://github.com/pc2ccs/pc2v9/wiki/Event-Feed-Module page has a whole section on the _non-gui_ mode of the event feeder.  That being said, to log exceptions to a logfile is fine for the --nogui mode.  But yes, a message/dialog of some sort would be nice in the event the SS Certificate can not be generated.

Reply:
do you want an exit after they hit the ok on that dialog?

Reply:
anyways, i have added that code.